### ISSUE TITLE
Add document and drawing number generation

### ISSUE THREAD

**Issue by stefi07** [uasal/documents_index#17]
# Summary
For change controlled entries of type "drawing", the "Number" field will guide the user through building the appropriate number for the entry according to the drawing tree structures.

For all entries of type "document", the "Number" field will autogenerate a alphanumeric string.

Entries with an Associated Number can also have additional Associated Handles (to support alternate numbering systems), but these would not be add-able or editable by the user.

A new "Assigned Numbers" page will show the list of already generated numbers.

### Number structure
Drawing numbers will follow the FOA & AOA trees as described in the [Drawing Tree Structure](https://gitlab.sc.ascendingnode.tech:8443/pearl-tele-design/payload/-/blob/develop/standards/naming.md#drawing-tree-structure) (GitLab link). The string elements of the number will be selected by the user as below and the numerical part of the number will be autogenerated (incrementally) to ensure uniqueness.

### Adding a new entry and generating a new number
1. Clicking the "Add Document" button (name needs updating) and setting the "Change controlled" field to "Yes" and the "Type" to "Drawing" opens the "Number" field
2. Each node in the drawing tree structure translates to a panel where the user chooses which branch of the tree to go down on.
3. Once the final choice is made, a "Confirm" button prompts the user to validate the generated number for this new controlled drawing
4. The user cannot save the new drawing entry without having confirmed the new number.
5. When the user clicks the save button on the new drawing both the entry and the new number are saved into the database and a link is created between them.

### Number structure
Document numbers will be represented by an alphanumeric string PRL-DOC-#####, where # is a number from 0-9. These numbers will be autogenerated (incrementally) to ensure uniqueness. All documents (change controlled or not) will have a Document Number.
To support existing alternate numbering systems (e.g. [halcyon](https://halcyon.as.arizona.edu/doclists/stpfoafulltable.html)), each Document Number could have additional Associated Handles. These cannot be added or edited by the user, but would be listed in the entry's '#' column and would retrieve the entry's details page.

### Adding a new entry and generating a new number
1. Clicking the "Add Document" button (name needs updating) and setting the "Change controlled" field to "Yes" and the "Type" to "Document" opens the "Number" field
2. On opening, the "Number" field is pre-populated with the next available document number.
3. When the user clicks the save button on the new drawing both the entry and the new number are saved into the database and a link is created between them.
* No user interface would be available for handles and they would need to be added directly in the database (to avoid validation rabbit-hole).

## Editing and existing entry
- When editing an existing document / drawing entry and setting the "Change controlled" field to "Yes" that will also open the "Number" field
- The rest of the workflow will follow steps 2-5 above for a Drawing or 2-3 for a Document.

## Existing Assigned Numbers
A new "Assigned Numbers" page will show the list of already generated numbers (grouped by type - "document" and "drawing") in a table with the columns:
1. Number
3. Associated Identifier (the Identifier of the associated document / drawing entry) and any Associated Handles
4. Title (of the associated document / drawing entry)
5. Author (of the associated document / drawing entry)
7. Maintained By (email of user responsible for the associated document / drawing entry)

## Properties
- Entries can be filtered by Assigned Number, Associated Handles, as well as by Identifier.
- An entry's Details page can be retrieved via its Assigned Number, Associated Handles, as well as its Identifier.
- An entry with Yes for Change Control can only be modified to "No" by Admin-level users.
- An entry with an Assigned Number can only be deleted by Admin-level users.
- An entry's Assigned Number cannot be removed.
- Existing Assigned Numbers cannot be deleted.

**Comment by sfrinaldi** [comment]
Heads up / can give you some more details in Slack but this is coming back to the surface for doing for teledocs. I believe its there just not active right now for the doc gen. 

We can talk about it next week though! 
@QuimbyCarson will be looped into this conversation for the part number portion / drawing numnbers.

**Comment by stefi07** [comment]
The updated version of [teledocs.space/demo](www.teledocs.space/demo) is now live.

It follows the diagram numbering format outlined in Carson's PowerPoint:
`(ESC/WCC)-(F/G/T)(A/P/X)ddd-dd` (where `d` signifies a digit)

You can view the numbers in use at [teledocs.space/demo/numbers](https://teledocs.space/demo/numbers)

The first set of three digits is generated automatically by incrementing the highest available number in the database for the corresponding combination of letters.

The second set of two digits (the configuration number) defaults to '00' for a new set of three digits. Users have the option to specify an existing set of three digits. In that case, the configuration number will be generated automatically by incrementing the highest available configuration number for that combination of letters and three digits.

If a user requests a set of three digits that doesn't exist already for that combination of letters, the form will propose the correct three digits in sequence.

I've added to the test database the existing reserved diagram numbers for ESC only.

### Notes and issues still open for decision:

- Change-controlled diagrams are currently un-editable for protection. Should we maintain this restriction and remove the Update button for such rows, or should we allow admins only to edit them?
- Change-controlled diagrams can currently only be deleted by admins, but the associated number is not deletable (it essentially remains reserved, as reassigning it to another drawing would cause confusion). Is it okay to keep it this way, or should the diagram entry be made non-deletable as well?
- Currently only change control drawings generate a number entry. Do we want this feature for type docs and/or type other too?
- I added the existing assigned numbers, but ESC-FA003, ESC-FA004, ESC-FA005 have a config notation that doesn't follow the stated structure (instead, it's 00-xx). What does this notation mean?

**Comment by sfrinaldi** [comment]
## 2026-02-04:
- Meetup today / demo for teledocs was run through and the open questions
- Using [#-database-discussion](https://departmentofastronomy.enterprise.slack.com/archives/C07QYB0DH9V) slack channel for quick conversations

### Questions/Answers
_List of answers to the questions mentioned above in the previous comment_

Q1: Change-controlled diagrams are currently un-editable for protection. Should we maintain this restriction and remove the Update button for such rows, or should we allow admins only to edit them?
A1: Defaulting to how teledocs entries are being treated right now (admins or the original owner will be the ones maintaining / in charge of this). Good for now.

Q2: Change-controlled diagrams can currently only be deleted by admins, but the associated number is not deletable (it essentially remains reserved, as reassigning it to another drawing would cause confusion). Is it okay to keep it this way, or should the diagram entry be made non-deletable as well?
A2: Should be okay right now (the number should remain reserved still / not reused) Will be on the owner for the entry. [Rough answer]

Q3: Currently only change control drawings generate a number entry. Do we want this feature for type docs and/or type other too?
A3: Drawings would be the main items / source URL can be used still for directory location for the other file types that are associated with the part number. (Could check in with Ewan maybe on this too)

Q4: I added the existing assigned numbers, but ESC-FA003, ESC-FA004, ESC-FA005 have a config notation that doesn't follow the stated structure (instead, it's 00-xx). What does this notation mean?
A4: Some of the ESC items are being adjusted still / more updates needed for part numbers / drawings (more items for ESC) 
@gwest2 @QuimbyCarson Did I get that correct? 

-------------------

### Action Items
- [x] @stefi07 Update Teledocs (moving the demo to be officially 'live' / viewable for everyone) for Monday morning (Irina's time)
- [x] @QuimbyCarson Will work with Fernando for adding Teledocs entries for WCC first 
  - Notify Irina when good to go 
- [x] @stefi07 To transfer ownership to Fernando for WCC entries (assuming their are under Carson in this state)
  - Will be Grant for ESC when we get to that part (later) 
- [x] Everyone- report issues in this iteration before ESC is added / pausing on this (larger set)
- [x] Check in with @douglase most likely if Q3/A3 is correct for assumption (just the teledocs entry for the drawing of the part to be used for the URL)
  - Good for now / for Q2-> revisit #18

**Comment by douglase** [comment]
for Q2, is it possible to recover the deleted information? making things not deletable but archiveable or "retired' might be better
for q3, only drawings is OK as long as it is understood it will apply to electronics design drawings not just PDM/solidworks drawings.

**Comment by sfrinaldi** [comment]
I believe we have a backup still in sync right now for teledocs. @stefi07 let me know if any of that has changed or if there were issues you have noticed (just in case) . At the moment there is no archive functionality. Maybe the quickest way to implement that too would actually just be a flag on the teledocs entry to be like a tag / check box for status if its archived or not unless that can be done without too much work with other items. 

I agree on the electronics drawings part too!

**Comment by stefi07** [comment]
Regarding recovering deleted info, that’s something we’ve discussed in the past and is captured in issue #18. The idea was to introduce a ‘Status’ column where we can deprecate / archive entries rather than delete them. If we’re happy with that idea I’ll have a look at it tomorrow and try and and roll it out with the rest on Monday (might be ambitious, but that'll be the aim at least :)).

Stefanie, you’re right that UITS keeps backups of our databases, so if anything goes particularly wrong, we can request them to reinstate one of the backups.

Regarding the electronics backups, would they follow the same naming structure? Otherwise I see no issues.

**Comment by sfrinaldi** [comment]
Sounds good to me! (We can continue with the roll out for Monday even if that part is not live for the status column but nice to see this was brought up before for another issue on this / seems like something good to have in the future)

Electronics- We might need to loop in @summerfranks in this to make sure or if they will be getting that part number once imported into solidworks (or something like that). @QuimbyCarson I think you might have already started a bit of that discussion with Summer or were planning on to? (If I'm remembering that correctly)

**Comment by sfrinaldi** [comment]
Adding this to here, quick overview for steps right now for doing this presently:

- Click `add document` from teledocs
- Select `drawing` for `Type`
- Select `Change Controlled` -> `yes`
  - (It should walk you through the new number process)

**Example screenshot:**
<img width="783" height="1328" alt="Image" src="https://github.com/user-attachments/assets/5b5ad80b-6899-4598-a8e2-c6c5d206ae21" />

**Comment by stefi07** [comment]
I would also note that the Title / Name field is required and should be unique.

**Comment by sfrinaldi** [comment]
@QuimbyCarson 
Just checking- for like `WCC-FP006-01-15` as an example, is the title name intended to be different from the part number?
<img width="731" height="333" alt="Image" src="https://github.com/user-attachments/assets/bedcff2c-a06b-48f9-a846-ea50c9a6e109" />

Do we want to have the indicator at the end for the configuration numbers for the teledocs entries actually?

Mainly asking as at least on the PDM side there is no additional -00 for the config in the naming (ex WCC-FP006-01-15 if correctly the issue mentioned at the end of this comment is just WCC-FP006 on PDM). There are configuration numbers associated with the drawing on the PDM viewer too that I think are different /  indicating separate things but correct me if I'm wrong!

Lately, bringer of bad news, if you do not have PDM / that network drive setup -> location does not map up directly to the PDM viewer. 

The prefix: `https://escpdm.as.arizona.edu/solidworkspdm` needs to be added to the file path locations for the entries for it to go to the web viewer location / workable for people without PDM. (Also if someone had PDM mapped to a different network drive letter)

**Comment by stefi07** [comment]
Yes, good catch on the link side, Stephanie! It does need to be a full url for most people with PDM login to be able to access it, since the `d/` maps to a specific local network / drive configuration. I can do a batch replace in the database to save manual updates if you or @QuimbyCarson can confirm the full address.

Also it looks like all the current entries have the same link (d/Steward_Obs_PDM/STP/LAZ/WCC/Flight) - is that specific enough? Generally the idea is to have a link to the location of that file or that file's source, so if there are more specific sub-folders within PDM that would be ideal.

@QuimbyCarson, I second Stephanie's question regarding the `WCC-FP006-01-15` format. I'm assuming it's a shortcut to not adding each different config entry? If that's the case, doesn't that create the issue that, in this case for example `WCC-FP006-05` is not registered as taken in teledocs, but it is actually taken in PDM, so we're losing the audit trail? If manually adding each of them is the issue, I'm happy to help on this front if you can share the list of entries that need to be added. Or, if not, maybe there's other functionality that we need to add? Let me know.

**Comment by summerfranks** [comment]
Can we have STP and LAZ options in the dropdown in addition to WCC and ESC since most of the EE teams items apply to both WCC & ESC?

Summer Franks (they/she)
Lead Electronic Engineer, Pearl/Lazuli Telescopes Instrument & AOA Team
UASAL / CAAO / Mountain Operations
University of Arizona Steward Observatory
Tucson, AZ 85721
+1 (520) 312-0089
________________________________
From: Irina Stefan ***@***.***>
Sent: Friday, February 13, 2026 5:51 AM
To: uasal/documents_index ***@***.***>
Cc: Franks, Summer - (summerfranks) ***@***.***>; Mention ***@***.***>
Subject: [EXT] Re: [uasal/documents_index] Add document and drawing number generation (Issue #17)

[https://avatars.githubusercontent.com/u/144726379?s=20&v=4]

External Email

________________________________
stefi07 left a comment (uasal/documents_index#17)<https://github.com/uasal/documents_index/issues/17#issuecomment-3897074367>

Yes, good catch on the link side, Stephanie! It does need to be a full url for most people with PDM login to be able to access it, since the d/ maps to a specific local network / drive configuration. I can do a batch replace in the database to save manual updates if you or @QuimbyCarson<https://github.com/QuimbyCarson> can confirm the full address.

Also it looks like all the current entries have the same link (d/Steward_Obs_PDM/STP/LAZ/WCC/Flight) - is that specific enough? Generally the idea is to have a link to the location of that file or that file's source, so if there are more specific sub-folders within PDM that would be ideal.

@QuimbyCarson<https://github.com/QuimbyCarson>, I second Stephanie's question regarding the WCC-FP006-01-15 format. I'm assuming it's a shortcut to not adding each different config entry? If that's the case, doesn't that create the issue that, in this case for example WCC-FP006-05 is not registered as taken in teledocs, but it is actually taken in PDM, so we're losing the audit trail? If manually adding each of them is the issue, I'm happy to help on this front if you can share the list of entries that need to be added. Or, if not, maybe there's other functionality that we need to add? Let me know.

**Comment by stefi07** [comment]
Hey, currently, the numbering format implemented is `(ESC/WCC)-(F/G/T)(A/P/X)ddd-dd (where d signifies a digit)`.

We can update the dropdown as needed, but before making any changes it would be safest to have a numbering format that:
- everyone agrees on
- is unlikely to change, and
- is documented. 

I would like to avoid frequent updates to the numbering format for audit purposed (having numbers generated that don't match the format anymore etc.).

**Comment by sfrinaldi** [comment]
I agree / good with this personally. Looks like this is how it is for electronics too per @summerfranks. @QuimbyCarson did this need to be brought up in the mechanical meeting or can you talk / pass this through Rob Stone for a clear?

**Comment by sfrinaldi** [comment]
@QuimbyCarson Was the WCC entries all good for having them move to Fernando for the maintainer for those entries? (Looked like it at least but checking)

@stefi07 
Checked in with him- its good to make Fernando the owner for the WCC drawing entries. There are some entries that need to be delated that were brought up for the range ones (can Slack Carson to confirm those)

**Comment by sfrinaldi** [comment]
<img width="1874" height="427" alt="Image" src="https://github.com/user-attachments/assets/03716383-e91c-411e-bed3-ed721aaa6840" />
Following entries: WCC-FP006-01, -02, -03 and -04.
Need to be deleted

**Comment by stefi07** [comment]
Deleted WCC-FP006-01, -02, -03 and -04 and updated the owner to fcoronado@arizona.edu.

**Comment by sfrinaldi** [comment]
[celebrate]     Rinaldi, Stephanie - (srinaldi) reacted to your message:
________________________________
From: Irina Stefan ***@***.***>
Sent: Tuesday, February 24, 2026 10:31:19 AM
To: uasal/documents_index ***@***.***>
Cc: Rinaldi, Stephanie - (srinaldi) ***@***.***>; Comment ***@***.***>
Subject: [EXT] Re: [uasal/documents_index] Add document and drawing number generation (Issue #17)

[https://avatars.githubusercontent.com/u/144726379?s=20&v=4]

External Email

________________________________
stefi07 left a comment (uasal/documents_index#17)<https://url.usb.m.mimecastprotect.com/s/J0AvCqAWOqfzplmAgFZfnIE5COq?domain=github.com>

Deleted WCC-FP006-01, -02, -03 and -04 and updated the owner to ***@***.******@***.***>.

**Comment by sfrinaldi** [comment]
I think for the initial purpose of this issue the drawing number generation is complete (and indicated that document gen we can figure out at a later point for any items and standards for that format). 

Closing out (before we use this issue too much past the initial intent). 
We will need to do the same process for ESC entries _however_ for the teledocs side for the implementation for this feature this issue is completed.
